### PR TITLE
rework daml_compile bazel rule

### PR DESCRIPTION
**Pull request by nickchapman-da**
Rework `daml_compile` bazel rule.

Avoiding `damlc compile` and `damlc package` commands (which we would like to deprecate), and replacing with `damlc build` together with a post `.dar` to `.dalf` extraction step in the couple of places where we actually want the `.dalf` for testing.

Also cleanup up the interface to `daml_compile`
- Removed `main_src`. It is now just listed with the other `srcs`
- The DAML project's `version` can be specified. (This is used in the `build-and-lint` tests, because the version number is relied on there)

Note: Expect CI to fail because the extract step is currently implemented using `unzip`. I'm going to fix this by coding in haskell.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: language-support/hs/bindings/BUILD.bazel [hunk 1]
```diff
@@ -36,13 +36,11 @@ da_haskell_library(
 daml_compile(
     name = "for-tests",
     srcs = glob(["test/daml/for-tests/*.daml"]),
-    main_src = "test/daml/for-tests/Main.daml",
 )
 
 daml_compile(
     name = "for-upload",
     srcs = glob(["test/daml/for-upload/ExtraModule.daml"]),
```
Comment (shayne-fletcher) [inline language-support/hs/bindings/BUILD.bazel]: Isn't this dangerous? If the file doesn't exist, then `glob` means we'll just get the empty list potentially obscuring an error?

File: language-support/ts/codegen/tests/BUILD.bazel [hunk 2]
```diff
@@ -7,12 +7,12 @@ load("@build_environment//:configuration.bzl", "sdk_version")
...
 # This rule builds a dar from the daml sources under the 'daml'
-# directory. It is referenced from the 'build-and-lint' test target that
+# directory. It is referenced from the 'build-and-lint-test' target that
 # follows.
 daml_compile(
-    name = "build-and-lint-1.0.0",
+    name = "build-and-lint",
     srcs = glob(["daml/**/*.daml"]),
-    main_src = "daml/Main.daml",
+    version = "1.0.0",
```
Comment (shayne-fletcher) [inline language-support/ts/codegen/tests/BUILD.bazel]: Why is this necessary? The daml.yaml specifies the version as 1.0.0 doesn't it?

File: language-support/ts/codegen/tests/BUILD.bazel [hunk 3]
```diff
@@ -79,7 +79,7 @@ sh_test(
         "$(location //:daml2ts)",
         "$(location //ledger/sandbox:sandbox-binary_deploy.jar)",
         "$(location //ledger-service/http-json:http-json-binary_deploy.jar)",
-        "$(location :build-and-lint-1.0.0.dar)",
+        "$(location :build-and-lint.dar)",
```
Comment (shayne-fletcher) [inline language-support/ts/codegen/tests/BUILD.bazel]: 🎉 !

File: language-support/ts/codegen/tests/BUILD.bazel (hunk 2 above)
Comment (nickchapman-da) [reply language-support/ts/codegen/tests/BUILD.bazel]: You rely on it. Nowhere else does.
I've got a mind to change the default to something unusual like 1.2.3

File: language-support/hs/bindings/BUILD.bazel (hunk 1 above)
Comment (nickchapman-da) [reply language-support/hs/bindings/BUILD.bazel]: good point. I'll raise an explicit error for `[]`

File: rules_daml/daml.bzl [hunk 4]
```diff
@@ -1,94 +1,283 @@
...
+    project_version = ctx.attr.project_version
+    input_dar = ctx.file.dar
+    output_dalf = ctx.outputs.dalf
+    ctx.actions.run_shell(
+        inputs = [input_dar],
+        outputs = [output_dalf],
+        progress_message = "Extract DALF from DAR (%s)" % project_name,
+        command = """
+            set -eou pipefail
+            unzip -q {input_dar}
```
Comment (aherrmann-da) [inline rules_daml/daml.bzl]: Bazel offers `@bazel_tools//tools/zip:zipper` which you can use instead of `unzip`. See [here](https://github.com/bazelbuild/bazel/blob/18bd1056b00e1fd58638d4db81e02cfb7df8ce69/third_party/ijar/zip_main.cc#L361).

File: language-support/hs/bindings/BUILD.bazel (hunk 1 above)
Comment (aherrmann-da) [reply language-support/hs/bindings/BUILD.bazel]: Bazel itself now errors on empty globs unless you explicitly allow them with [`allow_empty`](https://docs.bazel.build/versions/master/be/functions.html#glob).

File: language-support/ts/codegen/tests/BUILD.bazel (hunk 2 above)
Comment (shayne-fletcher) [reply language-support/ts/codegen/tests/BUILD.bazel]: Something is not right.
```
shaynefletcher:foo$ cat daml.yaml
sdk-version: 0.13.56-snapshot.20200318.3529.0.6ea118d6
name: foo
source: daml
init-script: Setup:initialize
parties:
  - Alice
  - Bob
version: 0.0.1
dependencies:
  - daml-prim
  - daml-stdlib
  - daml-script
```
If I now,
```
daml build
```
I get,
```
Created .daml/dist/foo-0.0.1.dar
```

Comment (shayne-fletcher) [reply language-support/ts/codegen/tests/BUILD.bazel]: Oh I see what you are getting at. Somehow this test is relying on 1.0.0. That's a mistake. I'll amend it to expect 0.0.1. If I do that will it remove the need for the version string in the invocation of the rule?

Comment (shayne-fletcher) [reply language-support/ts/codegen/tests/BUILD.bazel]: I got it now. Thank-you!

File: language-support/hs/bindings/BUILD.bazel (hunk 1 above)
Comment (cocreature) [reply language-support/hs/bindings/BUILD.bazel]: A DALF doesn’t need to have any modules. I think you just want to omit the glob and use the list directly here.

File: rules_daml/daml.bzl [hunk 5]
```diff
@@ -1,94 +1,283 @@
...
+
+def _daml_configure_impl(ctx):
+    project_name = ctx.attr.project_name
+    project_version = ctx.attr.project_version
+    daml_yaml = ctx.outputs.daml_yaml
+    target = ctx.attr.target
+    ctx.actions.write(
+        output = daml_yaml,
+        content = """
+            sdk-version: 0.0.0
```
Comment (cocreature) [inline rules_daml/daml.bzl]: Please use the actual SDK version here instead of `0.0.0`. This doesn’t matter too much for DARs used for tests but it does matter for DARs like the trigger library since `daml build` refuses to use `dependencies` with a different `sdk-version` field (for good reasons).

Given that we did actually just fix this for `damlc package` and therefore `daml_compile` rules in #4999, this would be a regression so let’s please fix this in this PR.

File: rules_daml/daml.bzl [hunk 6]
```diff
@@ -1,94 +1,283 @@
...
+
+_daml_build = rule(
+    implementation = _daml_build_impl,
+    attrs = {
+        "daml_yaml": attr.label(
+            allow_single_file = True,
+            mandatory = True,
+            doc = "The daml.yaml config file.",
+        ),
+        "damls": attr.label_list(
```
[the remark this answers is not in the document]
Comment (cocreature) [reply rules_daml/daml.bzl]: :+1:

File: language-support/hs/bindings/BUILD.bazel (hunk 1 above)
Comment (nickchapman-da) [reply language-support/hs/bindings/BUILD.bazel]: > Bazel itself now errors on empty globs unless you explicitly allow them with [`allow_empty`](https://docs.bazel.build/versions/master/be/functions.html#glob).

unfortunately the `allow_empty` defaults to True, so it's not very useful as a mistake catcher

I fixed the case where a glob wasn't necessary.

File: rules_daml/daml.bzl (hunk 6 above)
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply rules_daml/daml.bzl]: done

File: rules_daml/daml.bzl (hunk 4 above)
Comment (nickchapman-da) [reply rules_daml/daml.bzl]: @aherrmann-da  Thanks for this suggestion.
I changed from using `unzip` to `zipper x`. works great!
Unfortunately I think I may still have a problem because I am using `find`... is that available on windows?

(And a related question. Is `ln -s` possible on windows?)

Comment (cocreature) [reply rules_daml/daml.bzl]: > (And a related question. Is `ln -s` possible on windows?)

No, symlinks on windows are pretty much unusable.

I believe you can get `find` from `msys2`. The other `dev_env_tool` rules in `WORKSPACE` are probably a useful reference.

File: rules_daml/daml.bzl (hunk 5 above)
Comment (nickchapman-da) [reply rules_daml/daml.bzl]: How does it make any different what it is set to?
We are using `damlc` directly, and not going via the SDK/assistant.

Comment (cocreature) [reply rules_daml/daml.bzl]: The `sdk-version` that you set here is what ends up in the `MANIFEST.MF` in the DAR.

Comment (nickchapman-da) [reply rules_daml/daml.bzl]: ok, I fixed this now

File: rules_daml/daml.bzl (hunk 4 above)
Comment (nickchapman-da) [reply rules_daml/daml.bzl]: ok, I'm changed to getting `find` from the posix toolchain, as `scala.bzl` does
Lets see if it works...

Comment (nickchapman-da) [reply rules_daml/daml.bzl]: it did

Comment (cocreature) [reply rules_daml/daml.bzl]: Ah right the posix toolchain is much nicer.

File: rules_daml/daml.bzl [hunk 7]
```diff
@@ -1,94 +1,299 @@
...
+
+def _daml_configure_impl(ctx):
+    project_name = ctx.attr.project_name
+    project_version = ctx.attr.project_version
+    daml_yaml = ctx.outputs.daml_yaml
+    target = ctx.attr.target
+    ctx.actions.write(
+        output = daml_yaml,
+        content = """
+            sdk-version: {sdk}
```
Comment (cocreature) [inline rules_daml/daml.bzl]: This needs to be `ghc_version` not SDK version. `sdk_version` will break when you are on a snapshot release commit.

File: language-support/hs/bindings/BUILD.bazel (hunk 1 above)
Comment (aherrmann) [reply language-support/hs/bindings/BUILD.bazel]: > unfortunately the `allow_empty` defaults to True, so it's not very useful as a mistake catcher

Indeed, they've been pushing the switch to default to `allow_empty = False` to later and later times. Last I checked it should already have been switched in 2.0, but it looks like now it's scheduled for Bazel 4.0, see https://github.com/bazelbuild/bazel/issues/8195.

File: rules_daml/daml.bzl (hunk 7 above)
Comment (nickchapman-da) [reply rules_daml/daml.bzl]: Thanks. @garyverhaegen-da Is there doc to explain what versions shoul dbe used where?
I made the fix in PR #5094

Comment (cocreature) [reply rules_daml/daml.bzl]: `ghc_version` needs to be used for anything that ends up as the version of a GHC (DAML) package, e.g. for the version of `daml-stdlib`. `sdk-version` is the “official” way to refer to an SDK version e.g. in the assistant. In this case `sdk_version` is completely correct. Sorry for forgetting to send the comment. I keep getting confused by githubs UX.
